File: 0278-first-bad-version/0278-first-bad-version.rb
# The is_bad_version API is already defined for you.
# @param {Integer} version
# @return {boolean} whether the version is bad
# def is_bad_version(version):

# @param {Integer} n
# @return {Integer}

def first_bad_version(n)
    l, r = 1, n
    while l < r
        m = (l + r) / 2
        if r - l >= 2
            if is_bad_version(m)
                r = m
            else
                l = m
            end
        else
            if is_bad_version(m)
                return m
            else
                return m + 1
            end
        end
    end
    return l
end 

File: 0278-first-bad-version/0278-first-bad-version.kt
/* The isBadVersion API is defined in the parent class VersionControl.
      fun isBadVersion(version: Int) : Boolean {} */

class Solution: VersionControl() {
    override fun firstBadVersion(n: Int) : Int {
        var l : Int = 0;
        var r : Int = n;
        while(l <= r)
        {
            var m : Int = l + (r - l) / 2;
            if(isBadVersion(m))
                r = m - 1;
            else
                l = m + 1;
        }
        return l;
	}
}


/**


/* The isBadVersion API is defined in the parent class VersionControl.
      bool IsBadVersion(int version); */

public class Solution : VersionControl {
    public int FirstBadVersion(int n) {
        int l = 1, r = n;
        while(l <= r) 
        {
            int m = l + (r - l) / 2;
            if(IsBadVersion(m))
            {
                r = m - 1;
            }
            else
            {
                l = m + 1;
            }
        }
        return l;
    }
}

*/